src/wallet/coinselection.h

```diff
@@ -174,10 +174,15 @@ struct CoinSelectionParams {
      * 1) Received from other wallets, 2) replacing other txs, 3) that have been replaced.
      */
     bool m_include_unsafe_inputs = false;
+    /**
+     * The maximally sized transaction we are willing to make.
+     */
+    std::optional<int64_t> m_max_tx_weight{std::nullopt};
 
-    CoinSelectionParams(FastRandomContext& rng_fast, size_t change_output_size, size_t change_spend_size,
+    CoinSelectionParams(FastRandomContext& rng_fast, int change_output_size, size_t change_spend_size,
```

Comment:
Thanks for your review @murchandamus 
> Perhaps if we are getting big into the business of changing the types of various transaction size, vsize and weight variables, we should perhaps consider introducing explicit types for weight, size, and vsize to prevent people accidentally mistaking them for the same type.

I agree with this suggestion and have same concern previously here https://github.com/bitcoin/bitcoin/pull/29242#discussion_r1469871777 should make the code cleaner

Sorry for not explaining the rationale behind this in the commit messages, I will try and explain why I did the change below


On master  we have
```cpp
    int max_inputs_weight = MAX_STANDARD_TX_WEIGHT - (coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR);
    

   .......
    max_inputs_weight -= (coin_selection_params.change_output_size * WITNESS_SCALE_FACTOR);

```

After this PR maximum transaction weight can be configured by users, so we now subtract from user provided maximum weight or use the default if not given.
```cpp
int max_selection_weight = coin_selection_params.m_max_tx_weight.value_or(MAX_STANDARD_TX
_WEIGHT) - coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR;
```
 
 `m_max_tx_weight` is `int`,` tx_noinputs_size` and `change_output_size`  are both `size_t`.
given that [Bitcoin Core only supports platforms where int is exactly 32 bits](https://github.com/bitcoin/bitcoin/blob/v26.0/src/compat/assumptions.h#L44)
 For signed integers (int), it ranges from -2^(31) to (2^(31)) - 1.
 For unsigned integers (unsigned int), it ranges from 0 to (2^(32)) - 1.
 
The issue arrises when  after compiling with "ASan + LSan + UBSan + integer, without dependencies, USDT" and running the functional test.

From  854acd72e9 we allow passing 0 weight  and test for that in the functional test,  results of the subtracting `coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR` from 0 causes  `max_selection_weight`  to have some positive value instead of 0. hence the test fails see  https://cirrus-ci.com/task/6341256662482944, This is reproducible locally.

Thats why I changed  `change_output_size`,  and `tx_noinputs_size`from `size_t` to `int64_t` and also use same datatype for `m_max_tx_weight` which works fine.



```cpp
int64_t max_selection_weight = coin_selection_params.m_max_tx_weight.value_or(MAX_STANDARD_TX
_WEIGHT) - static_cast<int64_t>(coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR);
```

But since [Bitcoin Core only supports platforms where int is exactly 32 bits](https://github.com/bitcoin/bitcoin/blob/v26.0/src/compat/assumptions.h#L44) I think using int for `change_output_size`,  and `tx_noinputs_size` should work fine?

Reply:
> From https://github.com/bitcoin/bitcoin/commit/854acd72e99dc9f0924c5f0b5b7a657b39538911 we allow passing 0 weight and test for that in the functional test.

After thinking about this more I think it does not make sense to allow passing passing 0 maximum transaction weight, should instead be between minimum transaction weight possible and `MAX_STANDARD_TX_WEIGHT`

Reply:
I switched back to status quo, and prevent users from passing `max_tx_weight` below `MIN_STANDARD_TX_NONWITNESS_SIZE * WITNESS_SCALE_FACTOR`

Reply:
> But since [Bitcoin Core only supports platforms where int is exactly 32 bits](https://github.com/bitcoin/bitcoin/blob/v26.0/src/compat/assumptions.h#L44) I think using int for `change_output_size`, and `tx_noinputs_size` should work fine?

[…]

> I switched back to status quo, and prevent users from passing `max_tx_weight` below `MIN_STANDARD_TX_NONWITNESS_SIZE * WITNESS_SCALE_FACTOR`

I think we may have crossed some wires here. Thanks for the explanation, that was very helpful. I still agree that `size_t` is the _wrong_ type for `change_output_size` and `tx_noinputs_size`. I would support you taking the opportunity because `m_max_tx_weight` is `int` to also change those two to `int`, perhaps even take `change_spend_size` along for that. I would just prefer that it would be better to make it a separate commit and that it should be explained in the commit message.

Reply:
I Added a commit that changed them to `int` :)